charms/candid/src/charm.py

```diff
@@ -0,0 +1,323 @@
+        """
+        Install candid snap
+        """
+        self.set_status_and_log("Installing snap", WaitingStatus)
+        self._install_snap()
+        if self.snap_installed:
+            self.set_status_and_log("Snap installed", WaitingStatus)
+
+    def _start(self, event: StartEvent):
+        """
+        Starts candidsrv service
+        """
+        if self._check_config(event):
+            self.set_status_and_log("Starting candid", WaitingStatus)
+            self.snap.start(["candidsrv"])
+            if self.snap_running:
+                self.set_status_and_log("Ready ", ActiveStatus)
+
+    def _config_changed(self, event: StartEvent):
+        """
+        Updates snap internal configuration.
+        """
+        self._update_config_and_restart(event)
+
+    def _update_config_and_restart(self, event):
+        if not self._check_config(event):
+            return
+
+        config_values = {
+            key: value for key, value in self.config.items() if value
+        }
+        if self._stored.db_uri:
+            config_values["storage"] = {
+                "type": "postgres",
+                "connection-string": self._stored.db_uri,
+            }
+        else:
+            logging.info("db_uri not stored")
+
+        # convert identity-providers to json
+        # ids = config_values.get("identity-providers", "")
+        # if ids:
+        #    config_values["identity-providers"] = base64.b64encode(
+        #        ids.encode("utf-8")
+        #    ).decode("utf-8")
+
+        logging.info("setting config values {}".format(config_values))
+
+        config_values = flatten_dict(config_values)
+        config_values = {
+            "candid.{}".format(key): value
+            for key, value in config_values.items()
+            if value
+        }
+        try:
+            self.snap.set(config_values)
+        except Exception as e:
+            logging.error(
+                "error setting snap configuration values: {}".format(e)
+            )
+        self.set_status_and_log("Restarting.", WaitingStatus)
+        self.snap.restart(["candidsrv"])
+        if self.snap_running:
+            self.set_status_and_log("Ready", ActiveStatus)
+
+    def _check_config(self, event) -> bool:
+        """
+        Checks if required config is set and relations added.
+        """
+        if not self.snap_installed:
+            # TODO: We need error status, how?
+            self.set_status_and_log("Snap not installed", MaintenanceStatus)
+            return False
+
+        if self.model.get_relation(self.db.relation_name) is None:
+            self.set_status_and_log(
+                "Waiting for postgres relation.", BlockedStatus
+            )
+            event.defer()
+            return False
+
+        if self._stored.db_uri is None:
+            self.set_status_and_log(
+                "Waiting for postgres connection string", WaitingStatus
+            )
+            event.defer()
```

Comment:
Do we need to defer this hook? When pg is related this method will get called again because of the call to self._update_config_and_restart(event) on line 246

Reply:
defer should still work, i think